gui/freehand.py

```diff
@@ -657,9 +657,8 @@ def _process_queued_event(self, tdw, event_data):
             else:
                 drawstate.avgtime = (tavg, nevents)
 
-        # Refuse drawing if the layer is locked or hidden
         current_layer = model._layers.current
-        if current_layer.locked or not current_layer.visible:
+        if not current_layer.get_paintable():
```

Comment:
This recursive test might too much to do on every event, even in queues.

Also (and this is a more important shortcoming of the original code), the queued event handler should not be looking up the model's current layer on each queued event. Or even the target layer. What if the user has changed layers while there is a big queue to process? I think things are meant to flush before that happens, but it seems cleaner to avoid the *possiblity* of error too.

(So that's two problems... feel up for a fix of my code too?) :grinning: 

IMO, modes should observe the model more than they do right now. Painting modes should respond to potential changes of the get_paintable() state of the current layer by updating a boolean property. Use the boolean property at capture only, if capture is separate from painting. And FreehandMode should probably cache a target layer for its painting too, if that's possible.